### DISCUSSION TITLE
WhiteBoxTools input other than files?

### DISCUSSION THREAD

**Discussion by deeplook** [opengeos/leafmap#116] [Q&A] [answered]
I'm no expert in WhiteBoxTools at all, but after looking at a subset of the 200+ tools they appear to operate all on input _files_ that need to be present with some path in the filesystem. I wonder if this is not a bit of an inconvenience if you work with content you loaded from files or have in memory already and want to use WBT on? Or am I missing something?

**Comment by giswqs** [comment, marked as answer]
You are correct. The input files must be present in the file system. See the answer by the WhiteboxTools creator @jblindsay

> This is a feature that I have thought about in the past as well. Unfortunately there are some serious issues with it. First, I have developed WhiteboxTools to be front-end agnostic. Many of the R-front-end users would likely not be thrilled by the idea of modifying the application in a way that only allowed for certain features when interfacing with Python. Secondly, and as a consequence of the first point, WhiteboxTools actually only interacts with the front-ends through sub-process calls (front-end to back-end) and output calls (back-end to front-end). It is not a dynamic library (DLL), and therefore cannot share more complex data structures between Python and the application. Early on in the application's history I experimented with different design structures that would allow for this functionality, i.e. I toyed with the idea of developing WhiteboxTools as a dynamic library that would allow for a deeper integration with Python using FFI. However, the decision was eventually made to simplify the design by making WBT a command line tool that can be easily integrated into any front-end using system subprocess calls. As such, I'm afraid that this ship has sailed.

- https://github.com/jblindsay/whitebox-tools/issues/52#issuecomment-551162422
- https://github.com/jblindsay/whitebox-tools/issues/68
- https://github.com/giswqs/whiteboxR/issues/27
